steward/src/cellars/cellar_v2.rs

```
@@ -0,0 +1,702 @@
+        .params
+        .ok_or_else(|| sp_call_error("swap params cannot be unspecified".to_string()))?
+    {
+        UniV2OracleParams(p) => {
+            let mut path = Vec::<Token>::new();
+
+            for a in p.path {
+                let address = a.parse::<EthereumAddress>();
+                if address.is_err() {
+                    return Err(sp_call_error(format!(
+                        "could not parse swap params path address: {}",
+                        a
+                    )));
+                }
+
+                path.push(Token::Address(address.unwrap()))
+            }
+
+            let path = Token::Array(path);
+            Ok(abi::encode(&[path]))
+        }
+        UniV3OracleParams(p) => {
+            let mut path = Vec::<Token>::new();
+            for a in p.path {
+                let address = a.parse::<EthereumAddress>();
+                if address.is_err() {
+                    return Err(sp_call_error(format!(
+                        "could not parse swap params path address: {}",
+                        a
+                    )));
+                }
+
+                path.push(Token::Address(address.unwrap()))
+            }
+
+            let path = Token::Array(path);
+
+            let mut pool = Vec::<Token>::new();
+            for f in p.pool_fees {
+                pool.push(Token::Uint(f.into()))
+            }
+
+            let pool = Token::Array(pool);
+
+            Ok(abi::encode(&[path, pool]))
+        }
+    }
+}
+
+fn get_encoded_adaptor_call(data: Vec<AdaptorCall>) -> Result<Vec<AbiAdaptorCall>, Error> {
+    let mut result: Vec<AbiAdaptorCall> = Vec::new();
+    for d in data {
+        debug!("adaptor call to {}", d.adaptor);
+        let mut calls: Vec<Bytes> = Vec::new();
+        match d.call_data.unwrap() {
+            CallData::UniswapV3Calls(params) => {
+                for c in params.calls {
+                    match c.function.expect("adaptor function was empty") {
```

Similarly here, do we want to `expect` and panic if a value is missing? Don't protos have to have default zero values in the case of missing fields since "required" isn't really a concept in proto v3? I think in other places we use the `ok_or_else` method.